MapboxNavigation/NavigationMapView.swift

```diff
@@ -671,6 +676,188 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
+            
+            /**
+            Add to the count of remaining points all of the remaining legs.
+            */
+            for index in stride(from: routeProgress.legIndex + 1, to: completeRoutePoints.nestedList.count, by: 1) {
+                allRemainingPoints += completeRoutePoints.nestedList[index].flatMap{ $0 }.count
+            }
+            
+            /**
+            After calculating the number of remaining points and the number of all points,  calculate the index of the upcoming point.
+            */
+            let allPoints = completeRoutePoints.flatList.count
+            primaryRouteRemainingDistancesIndex = allPoints - allRemainingPoints - 1
+        } else {
+            primaryRouteRemainingDistancesIndex = nil
+        }
+    }
+    
+    private func getSlicedLinePointsCount(currentLegProgress: RouteLegProgress, currentStepProgress: RouteStepProgress) -> Int {
+        let startDistance = currentStepProgress.distanceTraveled
+        let stopDistance = currentStepProgress.step.distance
+        var slicedLine: Turf.LineString?
+        
+        /**
+        Implement the Turf.lineSliceAlong(lineString, startDistance, stopDistance) to return a sliced lineString.
+        */
+        if let lineString = currentStepProgress.step.shape {
+            if let midPoint = lineString.coordinateFromStart(distance: startDistance) {
+                slicedLine = lineString.trimmed(from: midPoint, distance: stopDistance - startDistance)
+            }
+        }
+        
+        if slicedLine != nil {
+            return slicedLine!.coordinates.count - 1
+        } else {
+            return 0
+        }
+    }
+
+    private func calculateRouteGranularDistances(coordinates: [CLLocationCoordinate2D]) -> RouteLineGranularDistances? {
+        if coordinates.isEmpty {
+            return nil
+        } else {
+            return calculateGranularDistances(points: coordinates)
+        }
+    }
+    
+    private func calculateGranularDistances(points: [CLLocationCoordinate2D]) -> RouteLineGranularDistances {
+        var distance = 0.0
+        var indexArray = [RouteLineDistancesIndex?](repeating: nil, count: points.count)
+        for index in stride(from: points.count - 1, to: 0, by: -1) {
+            let curr = points[index]
+            let prev = points[index - 1]
+            distance += calculateDistance(point1: curr, point2: prev)
+            indexArray[index - 1] = RouteLineDistancesIndex(point: prev, distanceRemaining: distance)
+        }
+        indexArray[points.count - 1] = RouteLineDistancesIndex(point: points[points.count - 1], distanceRemaining: 0.0)
+        return RouteLineGranularDistances(distance: distance, distanceArray: indexArray.compactMap{ $0 })
+    }
+
+    /**
+    Calculates the distance between 2 points using [EPSG:3857 projection](https://epsg.io/3857).
+    */
+    private func calculateDistance(point1: CLLocationCoordinate2D, point2: CLLocationCoordinate2D) -> Double {
```

Comment:
I think these calculation methods should be tested either directly of indirectly.